enterprise_gateway/services/sessions/kernelsessionmanager.py

```diff
@@ -279,22 +277,48 @@ def persistence_root_default(self):
     def __init__(self, kernel_manager, **kwargs):
         super(FileKernelSessionManager, self).__init__(kernel_manager, **kwargs)
         if self.enable_persistence:
-            self.kernel_session_file = os.path.join(self._get_sessions_loc(), KERNEL_SESSIONS_FILE_NAME)
-            self.log.info("Kernel session persistence location: {}".format(self.kernel_session_file))
+            self.log.info("Kernel session persistence location: {}".format(self._get_sessions_loc()))
 
     def save_sessions(self):
         if self.enable_persistence:
-            with open(self.kernel_session_file, 'w') as fp:
-                json.dump(KernelSessionManager.pre_save_transformation(self._sessions), fp)
-                fp.close()
+            os.rmdir(self._get_sessions_loc())
+            for kernel_id, session in self._sessions.items():
+                if kernel_id is not None:
+                    json_file_name = "".join(['kernel_', kernel_id, '.json'])
+                    kernel_session_file_path = os.path.join(self._get_sessions_loc(), json_file_name)
+                    temp_session = dict()
+                    temp_session[kernel_id] = session
+                    with open(kernel_session_file_path, 'w') as fp:
+                        json.dump(KernelSessionManager.pre_save_transformation(temp_session), fp)
+                        fp.close()
 
     def load_sessions(self):
         if self.enable_persistence:
-            if os.path.exists(self.kernel_session_file):
-                self.log.debug("Loading saved sessions from {}".format(self.kernel_session_file))
-                with open(self.kernel_session_file) as fp:
-                    self._sessions = KernelSessionManager.post_load_transformation(json.load(fp))
+            self._load_sessions()
+
+    def _load_sessions(self, kernel_id=None):
```

Comment:
You suggest we should have two more abstract methods load_session(id) save_session(id). Please correct me if I am wrong.

Reply:
Yes - every subclass will need to implement the singleton versions that load/save individual files/records/etc.

If we also add 'get_all' methods that return enumerations of the entries, then the full implementations for the plural methods (load_sessions, save_sessions) could be implemented in the abstract base class since there would no longer be "storage-specific" knowledge need for those methods.  But we can leave that for another exercise.